content/2017/02/09/generating-documentation-polyglot-projects.rst

```diff
@@ -0,0 +1,123 @@
+public: yes
+author: david
+tags: [
+  Docs,
+  Tools,
+  Javascript,
+  Python,
+  Sass
+  ]
+summary: |
+  Code documentation is ideally written as close to the actual code
+  as possible, but compiled into a comprehensive set of documentation
+  that includes code from all languages in use. Here's how we intend
+  to do that.
+
+
+Generating Code Documentation for Polyglot Projects
+===================================================
+
+At OddBird we believe that undocumented code is unfinished code.
+The act of documenting clarifies what we are building,
+and the resulting documentation guides consistency in how we do things
+in different parts of a project built by different team members
+at different times.
+
+In a typical project, one of our documentation deliverables is a
+living style guide which serves as documentation of the visual elements
+used in the app as well as guidelines for and examples of usage.
+The style guide is built automatically whenever we build the code, and
+it is delivered as a static HTML website which can be served alongside the app.
+For example, the style guide for this website can be found here:
+http://oddbird.net/styleguide/
+
+SassDoc
+-------
+
+We are currently building our style guides using a tool called `Sassdoc <http://sassdoc.com/>`_
+which compiles the documentation based on special comments written
+inline in our stylesheets. For example, the triple-slash commented lines
+in the following Sass:
+
+.. code:: scss
+
+  // Selection
+  // ---------
+  /// Selected text is highlighted in orange,
+  /// with any text-shadows removed.
+  /// @group typography
+  ::selection {
+    @include contrasted('select');
+    text-shadow: none;
+  }
+
+gets rendered in the style guide like this:
+
+.. image:: /static/images/blog/2017/docs/sassdoc.png
+
+We are working on our own theme for Sassdoc, called
+`Herman <https://github.com/oddbird/sassdoc-theme-herman/>`_,
+which provides extra tools for rendering samples of things like
+colors, fonts, and icons.
+
+The Multi-language Challenge
+----------------------------
+
+Generating documentation from inline comments like this is ideal
+for developer-oriented documentation (that is, documentation of
+how the system is built rather than how to use it). It keeps
+the documentation right next to the code that it documents,
+which both makes it more accessible and increases the chances
+that it will be kept up to date.
+
+But SassDoc only knows how to read comments from Sass.
+And the projects we work on implement patterns using multiple languages
+working together: Sass for stylesheets, Nunjucks/Jinja2 for markup,
+Javascript for interactivity, Python on the backend.
```

Comment:
Heh, only that in some of our projects we're using JavaScript for much more than "interactivity". E.g. in Qasper/Trig it's required to, ya know, view anything on the page. But I guess that's interactive? :)